### ISSUE TITLE
Clean Up Pods before Deleting Pod from API server

### ISSUE THREAD

**Issue by dashpole** [kubernetes/kubernetes#36466]
As discussed in #29033, we should only delete a pod after all of its resources have been reclaimed.
After speaking with @Random-Liu and @yujuhong, this is a larger issue, and one that should be discussed before proceeding.
cc: @derekwaynecarr @vishh @kubernetes/sig-node @kubernetes/sig-storage 

Here are some preliminary questions, and possible solutions, but I likely don't have the whole picture yet.

What kinds of things are currently not immediately deleted when we delete a pod (and are cleaned up via periodic housekeeping/garbage collection)?

- storage volumes/plugins
- Dead containers and their logs
- cgroups
- network interfaces

What methods can we use to trigger cleanup and verify that cleanup has completed?

- Just wait for garbage collection to kick in, and periodically check if it has completed (This is slow, probably not feasible, and breaks anything that relies on fast pod deletion)
- Adding the ability to “trigger” cleanup via a blocking function call (This increases code clutter, but is a viable solution)
- Using lifecycle events to trigger cleanup, and other events to show that cleanup has completed (This is a large change, but possibly the best solution long-term?)

**Comment by saad-ali** [comment]
> Which of these do we want to be deleted before notifying the API server when the pod is Terminated? Are any of them “not worth it”?
> * Volumes: Definitely need to be cleaned up before pod deletion

Kubelet volume manager currently monitors the pod manager and waits for a pod to be deleted as the signal to start volume cleanup. While, it could be modified to trigger off some other signal, containers running as part of the pod must always be terminated before volume cleanup is started otherwise there is a possibility of data corruption.

**Comment by vishh** [comment]
> What kinds of things are currently not immediately deleted when we delete a pod (and are cleaned up via periodic housekeeping/garbage collection)?

Dead containers & logs are one and the same as of now. 

> Which of these do we want to be deleted before notifying the API server when the pod is Terminated? Are any of them “not worth it”?

As mentioned above, logs are not an issue as of now. Image lifecycle is separate from that of pods since images are shared by default. Hence ignore images too for now.
I think all objects that belong to a pod consuming sizable cluster resources, be it cgroups, storage volumes/plugins, network interfaces, cluster level objects, etc. needs to be cleaned up as part of "deleting" a pod object on the API server. The "pod" object is the unique identifier for all these disparate resources being in-use.

> What methods can we use to trigger cleanup and verify that cleanup has completed?

> Just wait for garbage collection to kick in, and periodically check if it has completed (This is slow, probably not feasible, and breaks anything that relies on fast pod deletion)
This is possible if we can trigger garbage collection on demand and have it not rely on "deleted"/"non-existant" pod objects

> Adding the ability to “trigger” cleanup via a blocking function call (This increases code clutter, but is a viable solution)

Cleanup needs to happen prior to "pod object" deletion. Whether it has to "blocking" is questionable. I'd recommend running a separate reaper thread per-pod that is meant to attempt cleaning up all of a pod's resources after which it will delete the pod object.

> Using lifecycle events to trigger cleanup, and other events to show that cleanup has completed (This is a large change, but possibly the best solution long-term?)

Long term support should move towards a thread per-pod with well defined state machine. Let's not tackle that now. 

And can we move this discussion back to #29033?

**Comment by vishh** [comment]
@saad-ali 

> Kubelet volume manager currently monitors the pod manager and waits for a pod to be deleted as the signal to start volume cleanup. While, it could be modified to trigger off some other signal, containers running as part of the pod must always be terminated before volume cleanup is started otherwise there is a possibility of data corruption.

This is an important point. The state machine for a pod needs to delete containers prior to deleting "volumes".

**Comment by yujuhong** [comment]
> Cleanup needs to happen prior to "pod object" deletion. Whether it has to "blocking" is questionable. I'd recommend running a separate reaper thread per-pod that is meant to attempt cleaning up all of a pod's resources after which it will delete the pod object.

I think the pod worker can just do this. There is no need to create yet another thread for this.


> And can we move this discussion back to #29033?

#29033 was marked v1.5 and closed already. If we are going to discuss the internal implementation in kubelet, I'd rather start with a new issue.

**Comment by vishh** [comment]
> I think the pod worker can just do this. There is no need to create yet another thread for this.

Agreed!

> #29033 was marked v1.5 and closed already. If we are going to discuss the internal implementation in kubelet, I'd rather start with a new issue.

I re-opened #29033 since it is not yet "completed". I'd recommend creating a google doc to discuss internal design instead.

**Comment by dashpole** [comment]
Closed in favor of #29033

**Comment by yujuhong** [comment]
> I re-opened #29033 since it is not yet "completed". I'd recommend creating a google doc to discuss internal design instead.

That usually serves as a write-up or proposal *after* some discussion. I don't think using github for the initial discussion is bad.

**Comment by vishh** [comment]
I'm having trouble following discussions that span multiple issues. To
preserve context, at-least for this issue, it is better to move the
discussion back to #29033
<https://github.com/kubernetes/kubernetes/issues/29033>.
As for docs vs issues, I feel it's a personal preference. I find it hard to
follow details in an issue once it has more than 10-15 separate posts. For
issues to work for me, one (ideally the author) needs to keep the original
post up to date by capturing all other comments further down the issue,
which few people do in this project.
I find it much easier to evaluate design options via a google doc.
